apps/src/templates/sectionProgress/progressTables/ProgressTableDetailCell.jsx

```diff
@@ -0,0 +1,203 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import ReactDOM from 'react-dom';
+import {
+  levelTypeWithoutStatus,
+  studentLevelProgressType
+} from '@cdo/apps/templates/progress/progressTypes';
+import ProgressTableLevelBubble from './ProgressTableLevelBubble';
+import * as progressStyles from '@cdo/apps/templates/progress/progressStyles';
+import color from '@cdo/apps/util/color';
+import firehoseClient from '@cdo/apps/lib/util/firehose';
+import _ from 'lodash';
+import {LevelStatus} from '@cdo/apps/util/sharedConstants';
+
+const UNPLUGGED_BUBBLE_WIDTH = getUnpluggedWidth();
+
+const styles = {
+  container: {
+    ...progressStyles.flexBetween,
+    position: 'relative',
+    whiteSpace: 'nowrap'
+  },
+  background: {
+    height: 10,
+    backgroundColor: color.lighter_gray,
+    position: 'absolute',
+    left: 10,
+    right: 10
+  },
+  sublevelContainer: {
+    position: 'relative',
+    display: 'inline-block'
+  }
+};
+export default class ProgressTableDetailCell extends React.Component {
+  static widthForLevels(levels) {
+    return (
+      levels.reduce((sum, level) => {
+        return (
+          sum +
+          (level.isUnplugged
+            ? UNPLUGGED_BUBBLE_WIDTH
+            : progressStyles.BUBBLE_CONTAINER_WIDTH) +
+          (level.sublevels || []).length *
+            progressStyles.LETTER_BUBBLE_CONTAINER_WIDTH
+        );
+      }, 0) +
+      2 * progressStyles.CELL_PADDING
+    );
+  }
+
+  static propTypes = {
+    studentId: PropTypes.number.isRequired,
+    sectionId: PropTypes.number.isRequired,
+    levels: PropTypes.arrayOf(levelTypeWithoutStatus).isRequired,
+    studentProgress: PropTypes.objectOf(studentLevelProgressType).isRequired,
+    stageExtrasEnabled: PropTypes.bool
+  };
+
+  constructor(props) {
+    super(props);
+    this.recordBubbleClick = this.recordBubbleClick.bind(this);
+  }
+
+  shouldComponentUpdate(nextProps) {
+    return !_.isEqual(this.props, nextProps);
```

Comment:
Just curious, why do we need this?  What does this do differently from react's built-in mechanism?

Reply:
react's built-in mechanism returns `true`, unless your component extends `React.PureComponent`, in which case it does a shallow compare of all props (hence why `PureComponent` is only supposed to be used with simple props like strings, numbs, and bools). we do this same kind of check higher up in the component tree so in practice this shouldn't even be called, but since performance is crucial in this component, we want to make sure we don't do any unnecessary re-rendering. however since we can't leverage `PureComponent` in this case, we have to implement our own logic.

Reply:
Ah, I see.  _.isEqual is a deep comparison.  Thanks for the explanation!